fix: update isolation as new fixtures of certain scopes arrive (#2277)
ApeWorX/ape: docs/userguides/testing.md

It has the same interface as the [ChainManager](../methoddocs/managers.html#ape.managers.chain.ChainManager).

## Isolation

By default, tests run with chain-isolation.
This means, at the start of each test, a snapshot is taken.
After each test completes, the chain reverts to that snapshot from the beginning of the test.

By default, every `pytest` fixture is `function` scoped, meaning it will be replayed each time it is requested (no result-caching).
For example, if you deploy a contract in a function-scoped fixture, it will be re-deployed each time the fixture gets used in your tests.

> To only deploy once, you can use different scopes, such as `"session"`, `"package"`, `"module"`, and `"class"`.

Author: From pytest docs:

```
Fixture scopes
Fixtures are created when first requested by a test, and are destroyed based on their scope:

function: the default scope, the fixture is destroyed at the end of the test.
class: the fixture is destroyed during teardown of the last test in the class.
module: the fixture is destroyed during teardown of the last test in the module.
package: the fixture is destroyed during teardown of the last test in the package.
session: the fixture is destroyed at the end of the test session.
```

Imagine running pytest in a repo with multiple packages I guess.

Isolation also works on greater fixture scopes, `session`, `package`, `module`, and `class`.

> For example, if you define a session scope that deploys a contract and makes transactions, the state changes from those transactions remain in subsequent tests that request that fixture.

Reviewer: Something relating back to snapshots (unsure if correct)
```suggestion
For example, if you define a session scope that deploys a contract and makes transactions, the state changes from those transactions remain in subsequent tests that request that fixture thanks to a snapshot that is taken when that fixture is used.
```

In the following example, the `my_contract` fixture gets deployed upon it's first usage, which is the test `test_my_contract_0`.
During the test `test_something_else`, it may not have been deployed yet, as it was not requested, and it is defined before the other tests.
Then, during `test_my_contract_1`, instead of deploying again, it uses the cached result from the session-scoped fixture and the chain still has it in its state because the fixture is session-scoped and runs before the test-isolation.

```python
import pytest

@pytest.fixture(scope="session")
def my_contract(accounts, project):
    owner = accounts[0]
    contract = project.MyContract.deploy(sender=owner)
    # Can also do stateful transactions in a session-scoped fixture.
    contract.initialize(sender=owner)
    return contract

def test_something_else():
    ...

def test_my_contract_0(my_contract):
    my_contract.myMethod()

def test_my_contract_1(my_contract):
    my_contract.myMethod()
```

To disable isolation, run `ape test` with the `--disable-isolation` flag.
When isolation is disabled, the blockchain's state persist as the tests run.

> This will be more performant and less complicated, but will also cause you to have to state-cognizant in your tests.

Reviewer: ```suggestion
This will be more performant and less complicated, but will also cause non-deterministic results in your tests as each test inherits the state of whatever was run before it.
This may be further complicated when running with other pytest plugins such as `pytest-xdist` or `pytest-split` which re-arranges the order that tests are executed in (not recommended to use these plugins together with ape).
```

> This may be further complicated when running with other pytest plugins such as `pytest-xdist` or `pytest-split` which re-arranges the order that tests are executed in (not recommended to use these plugins together with ape).

Reviewer: ```suggestion
This may be further complicated when running with other pytest plugins such as `pytest-xdist` or `pytest-split` which re-arranges the order that tests are executed in (not recommended to use these plugins together with ape until specific support is added).
```
